staking/staking_hub.go

```diff
@@ -0,0 +1,409 @@
+		fmt.Sprintf("Stake increase:\t%s %s", address.String(), stakeBalance.String()),
+		logInfo,
+	)
+}
+
+// DecreaseStake decreases the account's staked balance if the account is present
+func (sh *StakingHub) DecreaseStake(address types.Address, unstakeBalance *big.Int) {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if sh.isStaker(address) && sh.StakingMap[address].Cmp(unstakeBalance) >= 0 {
+		sh.StakingMap[address] = big.NewInt(0).Sub(sh.StakingMap[address], unstakeBalance)
+
+		sh.log(
+			fmt.Sprintf("Stake decrease:\t%s %s", address.String(), unstakeBalance.String()),
+			logInfo,
+		)
+	}
+}
+
+// ResetStake resets the account's staked balance
+func (sh *StakingHub) ResetStake(address types.Address) {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if sh.isStaker(address) {
+		delete(sh.StakingMap, address)
+	}
+
+	sh.log(
+		fmt.Sprintf("Stake reset:\t%s", address.String()),
+		logInfo,
+	)
+}
+
+// GetStakedBalance returns an accounts staked balance if it is a staker.
+// Returns 0 if the address is not a staker
+func (sh *StakingHub) GetStakedBalance(address types.Address) *big.Int {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if sh.isStaker(address) {
+		return sh.StakingMap[address]
+	}
+
+	return big.NewInt(0)
+}
+
+// ComputeStakeAfterEvents goes over the pending events and returns the difference in value between staked and unstaked events
+// for the current address and block number
+func (sh *StakingHub) ComputeStakeAfterEvents(balance *big.Int, contextEvent PendingEvent) *big.Int {
+	stakedTally := balance
+
+	sh.EventQueueMutex.Lock()
+	defer sh.EventQueueMutex.Unlock()
+
+	// Go over events which have not yet been committed to find what will be
+	// the final staking tally for this address
+	for _, event := range sh.EventQueue {
+		// Check that we are working in the correct context
+		if event.BlockNumber != contextEvent.BlockNumber ||
+			event.Address.String() != contextEvent.Address.String() {
+			continue
+		}
+
+		if event.EventType == StakingEvent {
+			stakedTally.Add(stakedTally, event.Value)
+		} else if event.EventType == UnstakingEvent {
+			stakedTally = big.NewInt(0)
+		}
+	}
+
+	return stakedTally
+}
+
+// stakerMapping is a representation of a staked account balance
+type stakerMapping struct {
+	Address types.Address `json:"address"`
+	Stake   *big.Int      `json:"stake"`
+}
+
+// GetStakerMappings returns the staking addresses and their staking balances
+func (sh *StakingHub) getStakerMappings() []stakerMapping {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	mappings := make([]stakerMapping, len(sh.StakingMap))
+
+	var indx = 0
+	for address, stake := range sh.StakingMap {
+		mappings[indx] = stakerMapping{
+			Address: address,
+			Stake:   stake,
+		}
+		indx++
+	}
+
+	return mappings
+}
+
+// ClearEvents resets the event queue. Called from the consensus layer
+// to discard any events that might have been created during initial block creation
```

Comment:
Let callers decide on why they are calling this 🙂 
```suggestion
// ClearEvents resets the event queue
```

Reply:
Fixed 👍